In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import seaborn as sns

In [2]:
#leer el conjunto de datos
# En lenguaje R sería --> read.csv('housing_train.csv')
data = pd.read_csv('housing_train.csv')

In [3]:
#Poder ver 81 columnas en python
pd.options.display.max_columns = 81

In [10]:
#Ver las 5 primeras filas
# En lenguaje R sería --> head(data)
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [5]:
#Un resume de los datos numéricos
# En lenguaje R sería --> summary(data)
data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

        BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     46.549315   567.240411  1057.429452  1162.626712   346.992466   
std     161.319273   441.866955   438.705324   386.587738   436.528436   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   223.000000   795.750000   882.000000     0.000000   
50%       0.000000   477.500000   991.500000  1087.000000     0.000000   
75%       0.000000   808.000000  1298.250000  1391.250000   728.000000   
max    1474.000000  2336.000000  6110.000000  4692.000000  2065.000000   

       LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath  \
count   1460.000000  1460.000000   1460.000000   1460.000000  1460.000000   
mean       5.844521  1515.463699      0.425342      0.057534     1.565068   
std       48.623081   525.480383      0.518911      0.238753     0.550916   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1129.500000      0.000000      0.000000     1.000000   
50%        0.000000  1464.000000      0.000000      0.000000     2.000000   
75%        0.000000  1776.750000      1.000000      0.000000     2.000000   
max      572.000000  5642.000000      3.000000      2.000000     3.000000   

          HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  \
count  1460.000000   1460.000000   1460.000000   1460.000000  1460.000000   
mean      0.382877      2.866438      1.046575      6.517808     0.613014   
std       0.502885      0.815778      0.220338      1.625393     0.644666   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     14.000000     3.000000   

       GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  \
count  1379.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean   1978.506164     1.767123   472.980137    94.244521    46.660274   
std      24.689725     0.747315   213.804841   125.338794    66.256028   
min    1900.000000     0.000000     0.000000     0.000000     0.000000   
25%    1961.000000     1.000000   334.500000     0.000000     0.000000   
50%    1980.000000     2.000000   480.000000     0.000000    25.000000   
75%    2002.000000     2.000000   576.000000   168.000000    68.000000   
max    2

In [6]:
#Un resumen de todos los datos
data.describe(include = 'all')

Id   MSSubClass MSZoning  LotFrontage        LotArea Street  \
count   1460.000000  1460.000000     1460  1201.000000    1460.000000   1460   
unique          NaN          NaN        5          NaN            NaN      2   
top             NaN          NaN       RL          NaN            NaN   Pave   
freq            NaN          NaN     1151          NaN            NaN   1454   
mean     730.500000    56.897260      NaN    70.049958   10516.828082    NaN   
std      421.610009    42.300571      NaN    24.284752    9981.264932    NaN   
min        1.000000    20.000000      NaN    21.000000    1300.000000    NaN   
25%      365.750000    20.000000      NaN    59.000000    7553.500000    NaN   
50%      730.500000    50.000000      NaN    69.000000    9478.500000    NaN   
75%     1095.250000    70.000000      NaN    80.000000   11601.500000    NaN   
max     1460.000000   190.000000      NaN   313.000000  215245.000000    NaN   

       Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
count     91     1460        1460      1460      1460      1460         1460   
unique     2        4           4         2         5         3           25   
top     Grvl      Reg         Lvl    AllPub    Inside       Gtl        NAmes   
freq      50      925        1311      1459      1052      1382          225   
mean     NaN      NaN         NaN       NaN       NaN       NaN          NaN   
std      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
min      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
25%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
50%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
75%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
max      NaN      NaN         NaN       NaN       NaN       NaN          NaN   

       Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  \
count        1460       1460     1460       1460  1460.000000  1460.000000   
unique          9          8        5          8          NaN          NaN   
top          Norm       Norm     1Fam     1Story          NaN          NaN   
freq         1260       1445     1220        726          NaN          NaN   
mean          NaN        NaN      NaN        NaN     6.099315     5.575342   
std           NaN        NaN      NaN        NaN     1.382997     1.112799   
min           NaN        NaN      NaN        NaN     1.000000     1.000000   
25%           NaN        NaN      NaN        NaN     5.000000     5.000000   
50%           NaN        NaN      NaN        NaN     6.000000     5.000000   
75%           NaN        NaN      NaN        NaN     7.000000     6.000000   
max           NaN        NaN      NaN        NaN    10.000000     9.000000   

          YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd  \
count   1460.000000   1460.000000      1460     1460        1460        1460   
unique          NaN           NaN         6        8          15          16   
top             NaN           NaN     Gable  CompShg     VinylSd     VinylSd   
freq            NaN           NaN      1141     1434         515         504   
mean    1971.267808   1984.865753       NaN      NaN         NaN         NaN   
std       30.202904     20.645407       NaN      NaN         NaN         NaN   
min     1872.000000   1950.000000       NaN      NaN         NaN         NaN   
25%     1954.000000   1967.000000       NaN      NaN         NaN         NaN   
50%     1973.000000   1994.000000       NaN      NaN         NaN         NaN   
75%     2000.000000   2004.000000       NaN      NaN         NaN         NaN   
max     2010.000000   2010.000000       NaN      NaN         NaN         NaN   

       MasVnrType   MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
count        1452  1452.000000      1460      1460       1460     1423   
unique          4          NaN         4         

In [72]:
#Si quiero ver frecuencias por clase para una variable
data['MSZoning'].value_counts()

RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: MSZoning, dtype: int64

In [73]:
#Saber cuántos objetos de cada tipo hay
data.dtypes.value_counts() 

object     43
int64      35
float64     3
dtype: int64

In [7]:
#  En lenguaje R sería --> corr <- cor(data[,])
corr = data.corr()
corr

Id  MSSubClass  LotFrontage   LotArea  OverallQual  \
Id             1.000000    0.011156    -0.010601 -0.033226    -0.028365   
MSSubClass     0.011156    1.000000    -0.386347 -0.139781     0.032628   
LotFrontage   -0.010601   -0.386347     1.000000  0.426095     0.251646   
LotArea       -0.033226   -0.139781     0.426095  1.000000     0.105806   
OverallQual   -0.028365    0.032628     0.251646  0.105806     1.000000   
OverallCond    0.012609   -0.059316    -0.059213 -0.005636    -0.091932   
YearBuilt     -0.012713    0.027850     0.123349  0.014228     0.572323   
YearRemodAdd  -0.021998    0.040581     0.088866  0.013788     0.550684   
MasVnrArea    -0.050298    0.022936     0.193458  0.104160     0.411876   
BsmtFinSF1    -0.005024   -0.069836     0.233633  0.214103     0.239666   
BsmtFinSF2    -0.005968   -0.065649     0.049900  0.111170    -0.059119   
BsmtUnfSF     -0.007940   -0.140759     0.132644 -0.002618     0.308159   
TotalBsmtSF   -0.015415   -0.238518     0.392075  0.260833     0.537808   
1stFlrSF       0.010496   -0.251758     0.457181  0.299475     0.476224   
2ndFlrSF       0.005590    0.307886     0.080177  0.050986     0.295493   
LowQualFinSF  -0.044230    0.046474     0.038469  0.004779    -0.030429   
GrLivArea      0.008273    0.074853     0.402797  0.263116     0.593007   
BsmtFullBath   0.002289    0.003491     0.100949  0.158155     0.111098   
BsmtHalfBath  -0.020155   -0.002333    -0.007234  0.048046    -0.040150   
FullBath       0.005587    0.131608     0.198769  0.126031     0.550600   
HalfBath       0.006784    0.177354     0.053532  0.014259     0.273458   
BedroomAbvGr   0.037719   -0.023438     0.263170  0.119690     0.101676   
KitchenAbvGr   0.002951    0.281721    -0.006069 -0.017784    -0.183882   
TotRmsAbvGrd   0.027239    0.040380     0.352096  0.190015     0.427452   
Fireplaces    -0.019772   -0.045569     0.266639  0.271364     0.396765   
GarageYrBlt    0.000072    0.085072     0.070250 -0.024947     0.547766   
GarageCars     0.016570   -0.040110     0.285691  0.154871     0.600671   
GarageArea     0.017634   -0.098672     0.344997  0.180403     0.562022   
WoodDeckSF    -0.029643   -0.012579     0.088521  0.171698     0.238923   
OpenPorchSF   -0.000477   -0.006100     0.151972  0.084774     0.308819   
EnclosedPorch  0.002889   -0.012037     0.010700 -0.018340    -0.113937   
3SsnPorch     -0.046635   -0.043825     0.070029  0.020423     0.030371   
ScreenPorch    0.001330   -0.026030     0.041383  0.043160     0.064886   
PoolArea       0.057044    0.008283     0.206167  0.077672     0.065166   
MiscVal       -0.006242   -0.007683     0.003368  0.038068    -0.031406   
MoSold         0.021172   -0.013585     0.011200  0.001205     0.070815   
YrSold         0.000712   -0.021407     0.007450 -0.014261    -0.027347   
SalePrice     -0.021917   -0.084284     0.351799  0.263843     0.790982   

               OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  \
Id                0.012609  -0.012713     -0.021998   -0.050298   -0.005024   
MSSubClass       -0.059316   0.027850      0.040581    0.022936   -0.069836   
LotFrontage      -0.059213   0.123349      0.088866    0.193458    0.233633   
LotArea          -0.005636   0.014228      0.013788    0.104160    0.214103   
OverallQual      -0.091932   0.572323      0.550684    0.411876    0.239666   
OverallCond       1.000000  -0.375983      0.073741   -0.128101   -0.046231   
YearBuilt        -0.375983   1.000000      0.592855    0.315707    0.249503   
YearRemodAdd      0.073741   0.592855      1.000000    0.179618    0.128451   
MasVnrArea       -0.128101   0.315707      0.179618    1.000000    0.264736   
BsmtFinSF1       -0.046231   0.249503      0.128451    0.264736    1.000000   
BsmtFinSF2        0.040229  -0.049107     -0.067759   -0.072319   -0.050117   
BsmtUnfSF        -0.136841   0.149040      0.181133    0.114442   -0.495251   
TotalBsmtSF      -0.171098   0.391452      0.291066    0.363936    0

In [8]:
#   En lenguaje R sería --> corrGreater <- findCorrelation(corr, cutoff=0.8)
corrGreater = corr[corr>=.8]
corrGreater
#plt.figure(figsize=(12,8))
#sns.heatmap(corrGreater, cmap="Greens")

Id  MSSubClass  LotFrontage  LotArea  OverallQual  \
Id             1.0         NaN          NaN      NaN          NaN   
MSSubClass     NaN         1.0          NaN      NaN          NaN   
LotFrontage    NaN         NaN          1.0      NaN          NaN   
LotArea        NaN         NaN          NaN      1.0          NaN   
OverallQual    NaN         NaN          NaN      NaN          1.0   
OverallCond    NaN         NaN          NaN      NaN          NaN   
YearBuilt      NaN         NaN          NaN      NaN          NaN   
YearRemodAdd   NaN         NaN          NaN      NaN          NaN   
MasVnrArea     NaN         NaN          NaN      NaN          NaN   
BsmtFinSF1     NaN         NaN          NaN      NaN          NaN   
BsmtFinSF2     NaN         NaN          NaN      NaN          NaN   
BsmtUnfSF      NaN         NaN          NaN      NaN          NaN   
TotalBsmtSF    NaN         NaN          NaN      NaN          NaN   
1stFlrSF       NaN         NaN          NaN      NaN          NaN   
2ndFlrSF       NaN         NaN          NaN      NaN          NaN   
LowQualFinSF   NaN         NaN          NaN      NaN          NaN   
GrLivArea      NaN         NaN          NaN      NaN          NaN   
BsmtFullBath   NaN         NaN          NaN      NaN          NaN   
BsmtHalfBath   NaN         NaN          NaN      NaN          NaN   
FullBath       NaN         NaN          NaN      NaN          NaN   
HalfBath       NaN         NaN          NaN      NaN          NaN   
BedroomAbvGr   NaN         NaN          NaN      NaN          NaN   
KitchenAbvGr   NaN         NaN          NaN      NaN          NaN   
TotRmsAbvGrd   NaN         NaN          NaN      NaN          NaN   
Fireplaces     NaN         NaN          NaN      NaN          NaN   
GarageYrBlt    NaN         NaN          NaN      NaN          NaN   
GarageCars     NaN         NaN          NaN      NaN          NaN   
GarageArea     NaN         NaN          NaN      NaN          NaN   
WoodDeckSF     NaN         NaN          NaN      NaN          NaN   
OpenPorchSF    NaN         NaN          NaN      NaN          NaN   
EnclosedPorch  NaN         NaN          NaN      NaN          NaN   
3SsnPorch      NaN         NaN          NaN      NaN          NaN   
ScreenPorch    NaN         NaN          NaN      NaN          NaN   
PoolArea       NaN         NaN          NaN      NaN          NaN   
MiscVal        NaN         NaN          NaN      NaN          NaN   
MoSold         NaN         NaN          NaN      NaN          NaN   
YrSold         NaN         NaN          NaN      NaN          NaN   
SalePrice      NaN         NaN          NaN      NaN          NaN   

               OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  \
Id                     NaN        NaN           NaN         NaN         NaN   
MSSubClass             NaN        NaN           NaN         NaN         NaN   
LotFrontage            NaN        NaN           NaN         NaN         NaN   
LotArea                NaN        NaN           NaN         NaN         NaN   
OverallQual            NaN        NaN           NaN         NaN         NaN   
OverallCond            1.0        NaN           NaN         NaN         NaN   
YearBuilt              NaN   1.000000           NaN         NaN         NaN   
YearRemodAdd           NaN        NaN           1.0         NaN         NaN   
MasVnrArea             NaN        NaN           NaN         1.0         NaN   
BsmtFinSF1             NaN        NaN           NaN         NaN         1.0   
BsmtFinSF2             NaN        NaN           NaN         NaN         NaN   
BsmtUnfSF              NaN        NaN           NaN         NaN         NaN   
TotalBsmtSF            NaN        NaN           NaN         NaN         NaN   
1stFlrSF               NaN        NaN           NaN         NaN         NaN   
2ndFlrSF               NaN        NaN           NaN         NaN         NaN   
LowQualFinSF           NaN        NaN           NaN         

In [9]:
#  En lenguaje R sería --> cova = cov(data[,])
cov=data.cov()
cov['GarageYrBlt'] 

Id                    0.752557
MSSubClass           86.172938
LotFrontage          44.022081
LotArea           -6291.693689
OverallQual          18.200588
OverallCond          -8.655597
YearBuilt           598.923695
YearRemodAdd        324.207040
MasVnrArea         1151.316421
BsmtFinSF1         1741.991351
BsmtFinSF2         -357.071385
BsmtUnfSF          2089.072567
TotalBsmtSF        3473.992532
1stFlrSF           2230.674453
2ndFlrSF            768.704200
LowQualFinSF        -36.029134
GrLivArea          2963.349520
BsmtFullBath          1.580797
BsmtHalfBath         -0.455732
FullBath              6.556654
HalfBath              2.440954
BedroomAbvGr         -1.248805
KitchenAbvGr         -0.602196
TotRmsAbvGrd          5.813718
Fireplaces            0.746767
GarageYrBlt         609.582509
GarageCars            9.161174
GarageArea         2588.201757
WoodDeckSF          702.514599
OpenPorchSF         367.770066
EnclosedPorch      -443.898645
3SsnPorch            17.528531
ScreenPo

In [12]:
cov['TotalBsmtSF'] 

Id              -2.851118e+03
MSSubClass      -4.426302e+03
LotFrontage      4.268527e+03
LotArea          1.142145e+06
OverallQual      3.263034e+02
OverallCond     -8.352827e+01
YearBuilt        5.186807e+03
YearRemodAdd     2.636254e+03
MasVnrArea       2.887054e+04
BsmtFinSF1       1.045276e+05
BsmtFinSF2       7.417541e+03
BsmtUnfSF        8.051720e+04
TotalBsmtSF      1.924624e+05
1stFlrSF         1.389907e+05
2ndFlrSF        -3.342032e+04
LowQualFinSF    -7.091642e+02
GrLivArea        1.048612e+05
BsmtFullBath     6.996800e+01
BsmtHalfBath    -3.294995e-02
FullBath         7.824037e+01
HalfBath        -1.076701e+01
BedroomAbvGr     1.805534e+01
KitchenAbvGr    -6.660180e+00
TotRmsAbvGrd     2.036329e+02
Fireplaces       9.602237e+01
GarageYrBlt      3.473993e+03
GarageCars       1.424791e+02
GarageArea       4.564792e+04
WoodDeckSF       1.275796e+04
OpenPorchSF      7.187184e+03
EnclosedPorch   -2.560073e+03
3SsnPorch        4.808171e+02
ScreenPorch      2.066691e+03
PoolArea  

In [13]:
cov['GrLivArea'] 

Id               1.832811e+03
MSSubClass       1.663845e+03
LotFrontage      5.183670e+03
LotArea          1.380033e+06
OverallQual      4.309608e+02
OverallCond     -4.659665e+01
YearBuilt        3.158490e+03
YearRemodAdd     3.117808e+03
MasVnrArea       3.719917e+04
BsmtFinSF1       4.989250e+04
BsmtFinSF2      -8.171747e+02
BsmtUnfSF        5.578592e+04
TotalBsmtSF      1.048612e+05
1stFlrSF         1.149845e+05
2ndFlrSF         1.577039e+05
LowQualFinSF     3.441210e+03
GrLivArea        2.761296e+05
BsmtFullBath     9.499001e+00
BsmtHalfBath    -2.373510e+00
FullBath         1.823855e+02
HalfBath         1.098703e+02
BedroomAbvGr     2.234554e+02
KitchenAbvGr     1.158565e+01
TotRmsAbvGrd     7.050606e+02
Fireplaces       1.563982e+02
GarageYrBlt      2.963350e+03
GarageCars       1.834878e+02
GarageArea       5.269198e+04
WoodDeckSF       1.629669e+04
OpenPorchSF      1.149716e+04
EnclosedPorch    2.926882e+02
3SsnPorch        3.180224e+02
ScreenPorch      2.974197e+03
PoolArea  

In [14]:
cov['GarageArea'] 

Id               1.589549e+03
MSSubClass      -8.923920e+02
LotFrontage      1.853024e+03
LotArea          3.849872e+05
OverallQual      1.661850e+02
OverallCond     -3.605025e+01
YearBuilt        3.092857e+03
YearRemodAdd     1.640274e+03
MasVnrArea       1.446283e+04
BsmtFinSF1       2.895936e+04
BsmtFinSF2      -6.286505e+02
BsmtUnfSF        1.731721e+04
TotalBsmtSF      4.564792e+04
1stFlrSF         4.048257e+04
2ndFlrSF         1.291218e+04
LowQualFinSF    -7.027742e+02
GrLivArea        5.269198e+04
BsmtFullBath     1.988028e+01
BsmtHalfBath    -1.252455e+00
FullBath         4.778162e+01
HalfBath         1.758472e+01
BedroomAbvGr     1.138117e+01
KitchenAbvGr    -3.035400e+00
TotRmsAbvGrd     1.173989e+02
Fireplaces       3.709649e+01
GarageYrBlt      2.588202e+03
GarageCars       1.410015e+02
GarageArea       4.571251e+04
WoodDeckSF       6.020617e+03
OpenPorchSF      3.420130e+03
EnclosedPorch   -1.591326e+03
3SsnPorch        2.199300e+02
ScreenPorch      6.128902e+02
PoolArea  

## Análisis de covarianza

‘GarageYrBlt’- Este es el año en que se construyó el garaje de la casa.
‘YearBuilt’- Esta variable le da el año en que se realizó la construcción original de la propiedad.
‘TotalBsmtSF’- Como su nombre indica, esta métrica contiene el área total del sótano en pies cuadrados.
‘1stFlrSF’ y ‘2ndFlrSF’- Estas variables representan las áreas del primer y segundo piso (en pies cuadrados) respectivamente
‘GrLivArea’- Este es uno de los muchos parámetros disponibles relacionados con el área. Este indica específicamente el área de vivienda (en pies cuadrados) disponible en la casa.
‘TotRmsAbvGrd’- Este parámetro contiene el número total de habitaciones por encima del nivel (o suelo). El archivo de descripción de datos indica claramente que esto no tiene en cuenta los baños.
"GarageArea": ​​esta variable representa el tamaño del garaje en pies cuadrados.
‘GarageCars’- Esta columna contiene el tamaño del garaje en términos de capacidad del automóvil. Por lo tanto, todos los valores en esta métrica deben ser enteros.


GarageYrBlt YearBuilt =598.92
TotalBsmtSF 1stFlrSF = 1.389907e+05 138000
GrLivArea  TotRmsAbvGrd = 1.048612e+05 104800
GarageArea GarageCars = 1.410015e+02 141

## Justificar borrado de columnas
Ahora, en un caso, cuando las variables independientes están correlacionadas (esta condición es multicolinealidad), indica que los cambios en una variable también causarían cambios en otra variable. 
Cuanto más fuerte sea esta correlación, mayor será la dificultad de cambiar una variable sin cambiar otra. Esto hace que sea difícil para el modelo de regresión estimar la relación entre cada variable independiente y la variable dependiente de forma independiente, reduciendo así el rendimiento de su modelo de regresión. 


# Variables a eliminar
GarageArea
1stFlrSF 
GarageYrBlt
TotRmsAbvGrd

In [179]:
# En lenguaje R sería --> 
# borrar <- c('GarageYrBlt', 'TotRmsAbvGrd', 'GarageArea', '1stFlrSF')
# data <- data[ , !(names(data) %in% borrar)]

data = data.drop(['GarageYrBlt', 'TotRmsAbvGrd', 'GarageArea', '1stFlrSF'], axis = 1) 


In [180]:
#Saber cuántos datos no nulos hay
# En lenguaje R sería -->  is.null(data) or is.na(data)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 77 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [181]:
catFeature = data.columns[data.dtypes == object]
catFeature

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')

In [182]:
conFeature = data.columns[data.dtypes != object]
conFeature

Index(['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'Fireplaces', 'GarageCars',
       'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch',
       'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

In [183]:
data[catFeature].isna().sum().sort_values(ascending=True)

MSZoning            0
PavedDrive          0
Functional          0
KitchenQual         0
CentralAir          0
HeatingQC           0
Heating             0
SaleType            0
Foundation          0
ExterCond           0
ExterQual           0
Exterior2nd         0
Exterior1st         0
RoofMatl            0
SaleCondition       0
HouseStyle          0
Street              0
LotShape            0
RoofStyle           0
Utilities           0
LotConfig           0
LandContour         0
BldgType            0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
Electrical          1
MasVnrType          8
BsmtCond           37
BsmtQual           37
BsmtFinType1       37
BsmtExposure       38
BsmtFinType2       38
GarageType         81
GarageFinish       81
GarageQual         81
GarageCond         81
FireplaceQu       690
Fence            1179
Alley            1369
MiscFeature      1406
PoolQC           1453
dtype: int64

In [184]:
data = data.drop(['FireplaceQu', 'Fence', 'Alley', 'MiscFeature', 'PoolQC'], axis = 1) 

In [185]:
catFeature = data.columns[data.dtypes == object]
catFeature

Index(['MSZoning', 'Street', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond',
       'PavedDrive', 'SaleType', 'SaleCondition'],
      dtype='object')

In [186]:
data[catFeature].isna().sum().sort_values(ascending=False)

GarageCond       81
GarageQual       81
GarageFinish     81
GarageType       81
BsmtFinType2     38
BsmtExposure     38
BsmtCond         37
BsmtQual         37
BsmtFinType1     37
MasVnrType        8
Electrical        1
SaleCondition     0
BldgType          0
Condition2        0
Condition1        0
Neighborhood      0
LandSlope         0
RoofStyle         0
LotConfig         0
Utilities         0
LandContour       0
LotShape          0
Street            0
HouseStyle        0
ExterCond         0
RoofMatl          0
Exterior1st       0
Exterior2nd       0
ExterQual         0
SaleType          0
Foundation        0
Heating           0
HeatingQC         0
CentralAir        0
KitchenQual       0
Functional        0
PavedDrive        0
MSZoning          0
dtype: int64

## Reemplazo de variables

# 10 Variables categóricas. Reemplazar por la moda

GarageCond 
GarageQual
GarageFinish 
GarageType     
BsmtFinType2    
BsmtExposure
BsmtCond
BsmtQual
BsmtFinType1 
MasVnrType
Electrical 

In [187]:
#   # En lenguaje R sería --> 
# getmode <- function(v) {
#   uniqv <- unique(v)
#   uniqv[which.max(tabulate(match(v, uniqv)))]
# }
# i = 2
# data[is.na(data[,i]), i] <- getmode(data[,i], na.rm = TRUE)

maxGarageCond = data["GarageCond"].value_counts().idxmax()
data["GarageCond"].replace(np.nan, maxGarageCond, inplace=True)

maxGarageQual = data["GarageQual"].value_counts().idxmax()
data["GarageQual"].replace(np.nan, maxGarageQual, inplace=True)

maxGarageFinish = data["GarageFinish"].value_counts().idxmax()
data["GarageFinish"].replace(np.nan, maxGarageFinish, inplace=True)

maxGarageType = data["GarageType"].value_counts().idxmax()
data["GarageType"].replace(np.nan, maxGarageType, inplace=True)

maxBsmtFinType2 = data["BsmtFinType2"].value_counts().idxmax()
data["BsmtFinType2"].replace(np.nan, maxBsmtFinType2, inplace=True)

maxBsmtExposure = data["BsmtExposure"].value_counts().idxmax()
data["BsmtExposure"].replace(np.nan, maxBsmtExposure, inplace=True)

maxBsmtCond = data["BsmtCond"].value_counts().idxmax()
data["BsmtCond"].replace(np.nan, maxBsmtCond, inplace=True)

maxBsmtQual = data["BsmtQual"].value_counts().idxmax()
data["BsmtQual"].replace(np.nan, maxBsmtQual, inplace=True)

maxBsmtFinType1 = data["BsmtFinType1"].value_counts().idxmax()
data["BsmtFinType1"].replace(np.nan, maxBsmtFinType1, inplace=True)

maxMasVnrType = data["MasVnrType"].value_counts().idxmax()
data["MasVnrType"].replace(np.nan, maxMasVnrType, inplace=True)

maxElectrical = data["Electrical"].value_counts().idxmax()
data["Electrical"].replace(np.nan, maxElectrical, inplace=True)


In [188]:
data[catFeature].isna().sum().sort_values(ascending=False)

SaleCondition    0
Condition2       0
MasVnrType       0
Exterior2nd      0
Exterior1st      0
RoofMatl         0
RoofStyle        0
HouseStyle       0
BldgType         0
Condition1       0
SaleType         0
Neighborhood     0
LandSlope        0
LotConfig        0
Utilities        0
LandContour      0
LotShape         0
Street           0
ExterQual        0
ExterCond        0
Foundation       0
BsmtQual         0
PavedDrive       0
GarageCond       0
GarageQual       0
GarageFinish     0
GarageType       0
Functional       0
KitchenQual      0
Electrical       0
CentralAir       0
HeatingQC        0
Heating          0
BsmtFinType2     0
BsmtFinType1     0
BsmtExposure     0
BsmtCond         0
MSZoning         0
dtype: int64

In [189]:
data[conFeature].isna().sum().sort_values(ascending=False)

LotFrontage      259
MasVnrArea         8
SalePrice          0
LowQualFinSF       0
2ndFlrSF           0
TotalBsmtSF        0
BsmtUnfSF          0
BsmtFinSF2         0
BsmtFinSF1         0
YearRemodAdd       0
YrSold             0
YearBuilt          0
OverallCond        0
OverallQual        0
LotArea            0
MSSubClass         0
GrLivArea          0
BsmtFullBath       0
BsmtHalfBath       0
FullBath           0
HalfBath           0
BedroomAbvGr       0
KitchenAbvGr       0
Fireplaces         0
GarageCars         0
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
MiscVal            0
MoSold             0
Id                 0
dtype: int64

## Reemplazo de variables numéricas por la media
LotFrontage
https://www.kaggle.com/ogakulov/lotfrontage-fill-in-missing-values-house-prices

MasVnrArea




In [190]:
avgLotFrontage = data['LotFrontage'].astype('float').mean(axis=0)
data['LotFrontage'].replace(np.nan, avgLotFrontage, inplace=True)

avgLotFrontage
avgMasVnrArea = data['MasVnrArea'].astype('float').mean(axis=0)
data['MasVnrArea'].replace(np.nan, avgMasVnrArea, inplace=True)

In [191]:
data[conFeature].isna().sum().sort_values(ascending=True)

Id               0
MoSold           0
MiscVal          0
PoolArea         0
ScreenPorch      0
3SsnPorch        0
EnclosedPorch    0
OpenPorchSF      0
WoodDeckSF       0
GarageCars       0
Fireplaces       0
KitchenAbvGr     0
BedroomAbvGr     0
HalfBath         0
FullBath         0
BsmtHalfBath     0
BsmtFullBath     0
GrLivArea        0
MSSubClass       0
LotFrontage      0
LotArea          0
OverallQual      0
OverallCond      0
YearBuilt        0
YrSold           0
YearRemodAdd     0
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
2ndFlrSF         0
LowQualFinSF     0
MasVnrArea       0
SalePrice        0
dtype: int64

## Convertir variables categóricas en datos numéricos

In [192]:
catFeature = data.columns[data.dtypes == object]
catFeature
catFeature.tolist()

['MSZoning',
 'Street',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'SaleType',
 'SaleCondition']

In [193]:
# En lenguaje R sería --> 
# data$MSZoning<-factor(data$MSZoning)

#Aplicar codificador de etiquetas

from sklearn.preprocessing import LabelEncoder,OneHotEncoder

#LabelEncoder codifica las etiquetas de destino con un valor entre 0 y n_classes-1.

le = LabelEncoder()
#fit_transform --> Ajusta el codificador de etiquetas y devuelva las etiquetas codificadas
x = data[catFeature].apply(le.fit_transform)


#oneList=[]

#La entrada a este transformador debe ser una serie de enteros, que denota los valores adquiridos por las características categóricas (discretas). 
#Las características se codifican mediante un esquema de codificación de uno en caliente (también conocido como "uno de K" o "ficticio"). 
#Esto crea una columna binaria para cada categoría y devuelve una matriz dispersa o una matriz densa (dependiendo del parámetro disperso)

one = OneHotEncoder()
oneData = one.fit_transform(x).toarray()
oneData.shape

(1460, 234)

In [194]:
x

MSZoning  Street  LotShape  LandContour  Utilities  LotConfig  \
0            3       1         3            3          0          4   
1            3       1         3            3          0          2   
2            3       1         0            3          0          4   
3            3       1         0            3          0          0   
4            3       1         0            3          0          2   
...        ...     ...       ...          ...        ...        ...   
1455         3       1         3            3          0          4   
1456         3       1         3            3          0          4   
1457         3       1         3            3          0          4   
1458         3       1         3            3          0          4   
1459         3       1         3            3          0          4   

      LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  \
0             0             5           2           2         0           5   
1             0            24           1           2         0           2   
2             0             5           2           2         0           5   
3             0             6           2           2         0           5   
4             0            15           2           2         0           5   
...         ...           ...         ...         ...       ...         ...   
1455          0             8           2           2         0           5   
1456          0            14           2           2         0           2   
1457          0             6           2           2         0           5   
1458          0            12           2           2         0           2   
1459          0             7           2           2         0           2   

      RoofStyle  RoofMatl  Exterior1st  Exterior2nd  MasVnrType  ExterQual  \
0             1         1           12           13           1          2   
1             1         1            8            8           2          3   
2             1         1           12           13           1          2   
3             1         1           13           15           2          3   
4             1         1           12           13           1          2   
...         ...       ...          ...          ...         ...        ...   
1455          1         1           12           13           2          3   
1456          1         1            9           10           3          3   
1457          1         1            5            5           2          0   
1458          3         1            8            8           2          3   
1459          1         1            6            6           2          2   

      ExterCond  Foundation  BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  \
0             4           2         2         3             3             2   
1             4           1         2         3             1             0   
2             4           2         2         3             2             2   
3             4           0         3         1             3             0   
4             4           2         2         3             0             2   
...         ...         ...       ...       ...           ...           ...   
1455          4           2         2         3             3             5   
1456          4           1         2         3             3             0   
1457          2           4         3         1             3             2   
1458          4           1         3         3             2             2   
1459          4           1         3         3             3             1   

      BsmtFinType2  Heating  HeatingQC  CentralAir  Electrical  KitchenQual  \
0                5        1          0           1           4            2   
1                5        1          0           1           4            3   
2                5        1          0           1           4            2   
3 

In [195]:
#Borrar columna id 
data.drop('Id',inplace=True,axis=1)

## STANDARD SCALER
La idea detrás de StandardScaler es que transformará sus datos de modo que su distribución tenga un valor medio 0 y una desviación estándar de 1.
En el caso de datos multivariados, esto se hace en función de las características (en otras palabras, independientemente para cada columna de datos).
Dada la distribución de los datos, cada valor en el conjunto de datos tendrá el valor medio restado, y luego dividido por la desviación estándar de todo el conjunto de datos (o característica en el caso multivariante).

In [196]:
## En lenguaje R sería 

# dataNumeric = data %>% select(is.character)
# scaler = scale(dataNumeric)

#Reescalar columnas numéricas

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

#Columnas numéricas
conFeature = data.columns[data.dtypes != object]
col = conFeature.tolist()

#Aplicamos Standard Scaler 
dataNumeric = data[col]
sc = StandardScaler()
scaler = sc.fit(dataNumeric)
scaler.transform(dataNumeric)

array([[ 0.07337496, -0.22937175, -0.20714171, ..., -1.5991111 ,
         0.13877749,  0.34727322],
       [-0.87256276,  0.4519361 , -0.09188637, ..., -0.48911005,
        -0.61443862,  0.00728832],
       [ 0.07337496, -0.09311018,  0.07347998, ...,  0.99089135,
         0.13877749,  0.53615372],
       ...,
       [ 0.30985939, -0.18395123, -0.14781027, ..., -0.48911005,
         1.64520971,  1.07761115],
       [-0.87256276, -0.09311018, -0.08016039, ..., -0.8591104 ,
         1.64520971, -0.48852299],
       [-0.87256276,  0.22483348, -0.05811155, ..., -0.1191097 ,
         0.13877749, -0.42084081]])

In [197]:
#Creamos dataframe con columnas numéricas y categóricas codificadas
newData = np.append(dataNumeric,x,axis=1)
newData.shape

(1460, 71)

In [198]:
#Separar datos de entrenamiento y de test

xTrain,xTest,yTrain,yTest = train_test_split(newData,data['SalePrice'],test_size=0.2)

## Árboles de decisión problema de regresión

In [199]:
#Aplicar árboles de decisión 
# En lenguaje R sería
# library(tree)
#regressor <- tree(yTrain ~ . , xTrain)

from sklearn.tree import DecisionTreeRegressor

regressor = DecisionTreeRegressor(random_state=0)
regressor.fit(xTrain, yTrain)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=0, splitter='best')

In [200]:
predictValues = regressor.predict(xTest)

In [201]:
##Hallar Error cuádratico medio

from sklearn.metrics import mean_squared_error
mse = mean_squared_error(predictValues, yTest)
mse


33121384.188356165

## Random Forest problema de regresión

In [202]:
# En R sería 
# library(randomForest)
# randomForestModel <- randomForest(x = xTrain[, columnas], y = yTrain, data = xTrain, ntree = 500, do.trace= T)
                
from sklearn.ensemble import RandomForestRegressor

randomForestModel = RandomForestRegressor(n_jobs=2, n_estimators=500)
randomForestModel.fit(xTrain, yTrain)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=500, n_jobs=2, oob_score=False,
                      random_state=None, verbose=0, warm_start=False)

In [203]:
#Predicción
predictValuesRF=randomForestModel.predict(xTest)

In [204]:
mseRF = mean_squared_error(predictValuesRF, yTest)
mseRF

3208982.700109344

In [ ]:
#Baja el MSE con respecto a los árboles de decisión. RF mejor modelo. 

## Árboles de decisión, problema de clasificación

In [150]:
data['SalePrice']

0       208500
1       181500
2       223500
3       140000
4       250000
         ...  
1455    175000
1456    210000
1457    266500
1458    142125
1459    147500
Name: SalePrice, Length: 1460, dtype: int64

In [152]:
groupHousing = [
    (data['SalePrice'] <= 100000),
    (data['SalePrice'] > 100000) & (data['SalePrice'] <= 500000),
    (data['SalePrice'] > 500001)]
group = [1,2,3]

In [153]:
data['SalePrice_Grupo'] = np.select(groupHousing, group, default='null')

In [154]:
#Separar datos de entrenamiento y de test

xTrain,xTest,yTrain,yTest = train_test_split(newData,data['SalePrice_Grupo'],test_size=0.2)

In [155]:
from sklearn.tree import DecisionTreeClassifier

modelTree = DecisionTreeClassifier(random_state=0)
modelTree.fit(xTrain, yTrain)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [156]:
predictTree = modelTree.predict(xTest)

In [159]:
pd.crosstab(yTest, predictTree, rownames=["Real"], colnames=["Predicho"])

Predicho   1    2  3
Real                
1         18    0  0
2          1  270  0
3          0    1  2

## Ramdom Forest clasificación

In [161]:
from sklearn.ensemble import RandomForestClassifier

modelRFC = RandomForestClassifier(random_state=0)
modelRFC.fit(xTrain, yTrain)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [163]:
predictRFC = modelRFC.predict(xTest)

In [165]:
confusionMatrixRFC = pd.crosstab(yTest, predictRFC, rownames=["Real"], colnames=["Predicho"])
confusionMatrixRFC

Predicho   1    2  3
Real                
1         18    0  0
2          2  269  0
3          0    2  1